### ISSUE TITLE
pca error when apply it on test data

### ISSUE THREAD

**Issue by wwb306** [hyperopt/hyperopt-sklearn#104]
Through the training data, I got an estimator with PCA process, but when I made the prediction, I found some problems: when my test set number was different, I got different results.Further examination of the code revealed that the PCA process of the test set was refitted in the /hpsklearn/estimator.py file on 75 lines.Is there any problem? Under normal circumstances, the PCA process of the test set should be the reduced dimensional matrix generated by direct use of the training set PCA process.I can only comment out this line myself.

**Comment by adodge** [comment]
To paraphrase, the problem is that `HyperoptEstimator` fits the preprocessing as part of `transform_combine_XEX`, which is also called from `predict`.  So, it looks like we discard the trained preprocessing model when we use the HyperoptEstimator object to predict.

I'll admit, this combines two design decisions that I don't think I really understand (or maybe they just don't align with my use-case):

* Preprocessing as its own distinct input (I would rather use a pipeline and pass it as the `classifier` or `regressor`)
* The exogenous dataset stuff (I would rather do something like this http://scikit-learn.org/0.18/auto_examples/hetero_feature_union.html, which again ends up being just a pipeline)

The way I use hpsklearn results is, rather than keeping the HyperoptEstimator object, I extract the best preprocessor and classifier and put them in a pipeline.  This means I don't have to deploy hyperopt to production, which can be nice.  It also means I've never experienced this behavior.  (Which does seem wrong to me, but I'm not sure.)

I do something like this:

```
cls = HyperoptEstimator(preprocessing=PP, classifier=CLS)
cls = fit(X, y)
model = Pipeline([ ('pp%d'%i, p) for i,p in enumerate(cls._best_preprocs) ] + [ ('cls', cls._best_learner) ])

pred = model.predict(X1)
```

**Comment by wwb306** [comment]
Thank you. Now I use a transitional scheme, marking the training and prediction of the model separately, so that the fit PCA process is only used during the training process. The specific process is to change the following places in the `hpsklearn/estimator. Py` file:
1. Add a predictive tag is_predict in line 772 to skip the fit pca process:
`XEX = transform_combine_XEX(X, self.info, en_pps=self._best_preprocs, EXfit_list=EX_list, ex_pps_list= self._best_exprocs,is_predict=True)`

2. Also add one on line 60:
`Def transform_combine_XEX(Xfit, info, en_pps=[], Xval=None, EXfit_list=None, ex_pps_list=[], EXval_list=None,is_predict=False)`
3. Determine in line 75:
`If not is_predict:
    Pp_algo. Fit (Xfit)`
